### PR TITLE
Add default and latest stable edition to --edition in rustc

### LINKED ISSUE #106041: --edition should tell what the default edition is
rustc --edition: `error: Argument to option 'edition' missing`

rustc --edition=help: ``error: argument for `--edition` must be one of: 2015|2018|2021|2024. (instead was `help`)``

rustc --help

````
        --edition 2015|2018|2021|2024
                        Specify which edition of the compiler to use when
                        compiling code.
````

The default is not mentioned anywhere. I assume it is 2015?

### PR DESCRIPTION
**Pull request by sigaloid**
Fixes #106041

Boxes and leaks the formatted string as per https://github.com/rust-lang/rust/issues/106041#issuecomment-1364098669

It will now output the default edition and latest stable edition in the help message for the `--edition` flag.

### CODE REVIEW

**Comment by kadiwa4** [comment]
I think you can avoid one allocation by writing `Box::leak(format!(...).into_boxed_str())`

File: compiler/rustc_session/src/config.rs [hunk 1]
```diff
@@ -1339,7 +1339,8 @@ pub fn rustc_short_optgroups() -> Vec<RustcOptGroup> {
         opt::opt_s(
             "",
             "edition",
-            "Specify which edition of the compiler to use when compiling code.",
+            Box::leak(format!("Specify which edition of the compiler to use when compiling code.\
```
Comment (bjorn3) [inline compiler/rustc_session/src/config.rs:1342]: Could you please avoid this `Box::leak` call? It is possible to invoke rustc multiple times in the same process. Each time would leak memory. You should be able to inline the `opt_s` function to `pub fn opt_s(a: S, b: S, c: &str, d: S) -> R {` to allow strings with arbitrary lifetimes and then pass `&format(...)`.

Comment (sigaloid) [reply compiler/rustc_session/src/config.rs:1342]: Yep - I was actually pretty wary of this change but that's what I was recommended to do. I'll rewrite the type signature and see if it still works soon so we can avoid this leak.

Comment (sigaloid) [reply compiler/rustc_session/src/config.rs:1342]: Unfortunately I cannot refactor this out as the reference needs to outlive `'static`

```
error[E0521]: borrowed data escapes outside of function
    --> compiler/rustc_session/src/config.rs:1290:9
     |
1289 |     pub fn opt_s(a: S, b: S, c: &str, d: S) -> R {
     |                  -           -  - let's call the lifetime of this reference `'1`
     |                  |           |
     |                  |           `c` is a reference that is only valid in the function body
     |                  `a` declared here, outside of the function body
1290 |         stable(longer(a, b), move |opts| opts.optopt(a, b, c, d))
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |         |
     |         `c` escapes the function body here
     |         argument requires that `'1` must outlive `'static`
```

There's got to be something I'm missing - the enum variant for the formatted string are all static and embedded, and there's got to be a better option than either allocating, or just manually writing in the strings and commenting above the original variables, "this must be updated in xyz string". I don't know how but I think there might be a way to construct this statically somehow - it's just concatenating all-static strings.

Comment (bjorn3) [reply compiler/rustc_session/src/config.rs:1342]: I see. Another option would be to use `std::sync::LazyLock` to compute the description. If you put it in a static it should allow you to get a single `&'static str` without leaking anything.

**Comment by petrochenkov** [comment]
r? rust-lang/compiler
I'd like to take some break from reviewing.

**Comment by Nilstrieb** [comment]
Thank you for your PR! But I see you accidentally added a merge commit (probably by updating your branch with the github UI). This repository doesn't allow merge commit and you get rid of it. There's an explanation [in the dev guide](necessarily associated with failure), feel free to ask further questions if that doesn't help.

As for the `Box::leak`, I would prefer the solution suggested by bjorn3 to use a `LazyLock` instead of leaking the string. While leaking a few bytes here isn't a problem memory wise, we should still leak as few bytes as possible to help leak detectors.

**Comment by sigaloid** [comment]
Thanks for the recommendation! 

In regards to the LazyLock, are you recommending creating a static LazyLock<String> at the top of that function and referencing it in that function? I attempted it and could not get it to become a `&'static str` despite my best efforts.

**Comment by bjorn3** [comment]
Did you try `&*MY_STATIC`?

**Comment by sigaloid** [comment]
That worked @bjorn3 - thanks!

**Pull request closed without merging**